File: mobile/lib/core/providers/rocket_adapter.dart
import 'payment_models.dart';
import 'provider_adapter.dart';
import 'provider_id.dart';

/// Mock Rocket adapter — DBBL has no public developer API, so this
/// adapter returns deterministic mocks until an onboarding contract is
/// in place (RISK_REGISTER.md R-PRV-02).
class RocketAdapter implements ProviderAdapter {
  @override
  ProviderId get id => ProviderId.rocket;

  @override
  Future<PaymentInitiation> initiate(PaymentRequest req) async {
    if (req.type == 'MERCHANT_PAYMENT') {
      return PaymentInitiation(
        providerTxnId: 'ROCKET-M-${DateTime.now().millisecondsSinceEpoch}',
        redirectUrl: 'https://sandbox.rocket.example.com/mock?id=${req.idempotencyKey}',
        status: 'INITIATED',
      );
    }
    return launchDialerPassThrough(req);
  }

  @override
  Future<PaymentInitiation> launchDialerPassThrough(PaymentRequest req) async {
    final taka = req.amountMinorUnits ~/ 100;
    final ussd = '*322*1*${req.recipientPhone ?? ''}*$taka#';
    return PaymentInitiation(
      providerTxnId: 'ROCKET-D-${DateTime.now().millisecondsSinceEpoch}',
      status: 'INITIATED',
      instructions: PaymentInstructions(method: 'DIALER_PASS_THROUGH', ussdString: ussd),
    );
  }

  @override
  Future<PaymentStatus> pollStatus(String providerTxnId) async => PaymentStatus(
        providerTxnId: providerTxnId,
        status: 'SUCCESS',
        updatedAt: DateTime.now().toIso8601String(),
      );

  @override
  Future<PaymentReceipt> fetchReceipt(String providerTxnId) async => PaymentReceipt(
        providerTxnId: providerTxnId,
        providerId: ProviderId.rocket,
        type: 'MERCHANT_PAYMENT',
        status: 'SUCCESS',
        amountMinorUnits: 0,
        currency: 'BDT',
      );
}

File: backend/src/modules/payments/providers/rocket.adapter.ts
import { randomUUID } from 'node:crypto';

import { mockInitiation, mockReceiptFromRequest, mockStatus } from './mock.helpers.js';
import type {
  PaymentInitiation,
  PaymentReceipt,
  PaymentRequest,
  PaymentStatus,
  ProviderAdapter,
} from './provider.interface.js';

/**
 * Mock Rocket adapter.
 *
 * DBBL has not published a public developer API. This adapter returns
 * deterministic mock responses. See RISK_REGISTER.md R-PRV-02.
 */
export class RocketAdapter implements ProviderAdapter {
  readonly id = 'rocket' as const;

  async initiate(req: PaymentRequest): Promise<PaymentInitiation> {
    if (req.type === 'MERCHANT_PAYMENT') {
      return mockInitiation(req, {
        providerTxnId: `ROCKET-M-${randomUUID()}`,
        redirectUrl: `https://sandbox.rocket.example.com/mock/checkout?id=${req.idempotencyKey}`,
      });
    }
    return this.launchDialerPassThrough(req);
  }

  launchDialerPassThrough(req: PaymentRequest): Promise<PaymentInitiation> {
    const ussd = `*322*1*${req.recipientPhone ?? ''}*${Math.floor(req.amountMinorUnits / 100)}#`;
    return Promise.resolve(
      mockInitiation(req, {
        providerTxnId: `ROCKET-D-${randomUUID()}`,
        ussdString: ussd,
      }),
    );
  }

  pollStatus(providerTxnId: string): Promise<PaymentStatus> {
    return Promise.resolve(mockStatus(providerTxnId, 'SUCCESS'));
  }

  fetchReceipt(providerTxnId: string): Promise<PaymentReceipt> {
    return Promise.resolve(
      mockReceiptFromRequest(
        {
          providerId: 'rocket',
          type: 'MERCHANT_PAYMENT',
          amountMinorUnits: 0,
          currency: 'BDT',
          idempotencyKey: '',
        },
        providerTxnId,
      ),
    );
  }
}
